internal/cmd/grpc.go

```diff
@@ -0,0 +1,330 @@
+		return nil, err
+	}
+
+	server.pushShutdown(authShutdown)
+
+	// forward internal gRPC logging to zap
+	grpcLogLevel, err := zapcore.ParseLevel(cfg.Log.GRPCLevel)
+	if err != nil {
+		return nil, fmt.Errorf("parsing grpc log level (%q): %w", cfg.Log.GRPCLevel, err)
+	}
+
+	grpc_zap.ReplaceGrpcLoggerV2(logger.WithOptions(zap.IncreaseLevel(grpcLogLevel)))
+
+	// base observability inteceptors
+	interceptors := append([]grpc.UnaryServerInterceptor{
+		grpc_recovery.UnaryServerInterceptor(),
+		grpc_ctxtags.UnaryServerInterceptor(),
+		grpc_zap.UnaryServerInterceptor(logger),
+		grpc_prometheus.UnaryServerInterceptor,
+		otelgrpc.UnaryServerInterceptor(),
+	},
+		append(authInterceptors,
+			middlewaregrpc.ErrorUnaryInterceptor,
+			middlewaregrpc.ValidationUnaryInterceptor,
+			middlewaregrpc.EvaluationUnaryInterceptor,
+		)...,
+	)
+
+	if cfg.Cache.Enabled {
+		var cacher cache.Cacher
+
+		switch cfg.Cache.Backend {
+		case config.CacheMemory:
+			cacher = memory.NewCache(cfg.Cache)
+		case config.CacheRedis:
+			rdb := goredis.NewClient(&goredis.Options{
+				Addr:     fmt.Sprintf("%s:%d", cfg.Cache.Redis.Host, cfg.Cache.Redis.Port),
+				Password: cfg.Cache.Redis.Password,
+				DB:       cfg.Cache.Redis.DB,
+			})
+
+			server.pushShutdown(func(ctx context.Context) error {
+				return rdb.Shutdown(ctx).Err()
+			})
+
+			status := rdb.Ping(ctx)
+			if status == nil {
+				return nil, errors.New("connecting to redis: no status")
+			}
+
+			if status.Err() != nil {
+				return nil, fmt.Errorf("connecting to redis: %w", status.Err())
+			}
+
+			cacher = redis.NewCache(cfg.Cache, goredis_cache.New(&goredis_cache.Options{
+				Redis: rdb,
+			}))
+		}
+
+		interceptors = append(interceptors, middlewaregrpc.CacheUnaryInterceptor(cacher, logger))
+
+		logger.Debug("cache enabled", zap.Stringer("backend", cacher))
+	}
+
+	grpcOpts := []grpc.ServerOption{grpc_middleware.WithUnaryServerChain(interceptors...)}
+
+	if cfg.Server.Protocol == config.HTTPS {
+		creds, err := credentials.NewServerTLSFromFile(cfg.Server.CertFile, cfg.Server.CertKey)
+		if err != nil {
+			return nil, fmt.Errorf("loading TLS credentials: %w", err)
+		}
+
+		grpcOpts = append(grpcOpts, grpc.Creds(creds))
+	}
+
+	// initialize server
+	register.Add(fliptserver.New(logger, store))
+
+	// initialize grpc server
+	server.Server = grpc.NewServer(grpcOpts...)
+
+	// register grpcServer graceful stop on shutdown
+	server.pushShutdown(func(context.Context) error {
+		logger.Info("shutting down grpc server...")
+
+		server.Server.GracefulStop()
+
+		return nil
+	})
+
+	// register each grpc service onto the grpc server
+	register.RegisterGRPC(server.Server)
+
+	grpc_prometheus.EnableHandlingTimeHistogram()
+	grpc_prometheus.Register(server.Server)
+	reflection.Register(server.Server)
+
+	return server, nil
+}
+
+// ClientConn constructs and configures a client connection to the underlying gRPC server.
+func (s *GRPCServer) ClientConn(ctx context.Context) (*grpc.ClientConn, error) {
```

Comment:
since this doesnt actually depend on anything re: GRPCServer state, what do you think about moving this back to main inline? From what I can tell this method only relies on the `config` (`cfg`)

Reply:
this would also bring its method signature inline with the `httpServer` I think